Source/WebCore/animation/DeclarativeAnimation.cpp

```diff
@@ -318,33 +318,35 @@ void DeclarativeAnimation::invalidateDOMEvents(Seconds elapsedTime)
         } else if ((!wasIdle && !wasAfter) && isIdle)
             enqueueDOMEvent(eventNames().animationcancelEvent, elapsedTime, elapsedTime);
     } else if (is<CSSTransition>(this)) {
-        // https://drafts.csswg.org/css-transitions-2/#transition-events
-        if (wasIdle && (isPending || isBefore))
-            enqueueDOMEvent(eventNames().transitionrunEvent, intervalStart, effectTimeAtStart());
-        else if (wasIdle && isActive) {
-            auto scheduledEffectTime = effectTimeAtStart();
-            enqueueDOMEvent(eventNames().transitionrunEvent, intervalStart, scheduledEffectTime);
-            enqueueDOMEvent(eventNames().transitionstartEvent, intervalStart, scheduledEffectTime);
-        } else if (wasIdle && isAfter) {
-            enqueueDOMEvent(eventNames().transitionrunEvent, intervalStart, effectTimeAtStart());
-            enqueueDOMEvent(eventNames().transitionstartEvent, intervalStart, effectTimeAtStart());
-            enqueueDOMEvent(eventNames().transitionendEvent, intervalEnd, effectTimeAtEnd());
-        } else if ((m_wasPending || wasBefore) && isActive)
-            enqueueDOMEvent(eventNames().transitionstartEvent, intervalStart, effectTimeAtStart());
-        else if ((m_wasPending || wasBefore) && isAfter) {
-            enqueueDOMEvent(eventNames().transitionstartEvent, intervalStart, effectTimeAtStart());
-            enqueueDOMEvent(eventNames().transitionendEvent, intervalEnd, effectTimeAtEnd());
-        } else if (wasActive && isAfter)
-            enqueueDOMEvent(eventNames().transitionendEvent, intervalEnd, effectTimeAtEnd());
-        else if (wasActive && isBefore)
-            enqueueDOMEvent(eventNames().transitionendEvent, intervalStart, effectTimeAtEnd());
-        else if (wasAfter && isActive)
-            enqueueDOMEvent(eventNames().transitionstartEvent, intervalEnd, effectTimeAtStart());
-        else if (wasAfter && isBefore) {
-            enqueueDOMEvent(eventNames().transitionstartEvent, intervalEnd, effectTimeAtStart());
-            enqueueDOMEvent(eventNames().transitionendEvent, intervalStart, effectTimeAtEnd());
-        } else if ((!wasIdle && !wasAfter) && isIdle)
-            enqueueDOMEvent(eventNames().transitioncancelEvent, elapsedTime, elapsedTime);
+        if (m_owningElement->document().hasListenerType(Document::TRANSITION_ANIMATION_LISTENER)) {
```

Comment:
Maybe add this condition to the `else if (is<CSSTransition>(this))` statement to avoid a single, big nested `if`?